### DISCUSSION TITLE
10000 cycles of cyclic loading

### DISCUSSION THREAD

**Discussion by Ming-is** [neperfepx/FEPX#26] [Q&A] [answered]
I'm trying to use the restart module of fepx to simulate 10000 cycles of cyclic loading, but I can only calculate one cycle at a time. It seems that I can only manually add the cycle N file to simulate the cycles, which is too inefficient.
Does fepx have any instructions or methods to help quickly simulate a large number of cycles of cyclic loading
I look forward to your reply,
thanks

**Comment by acmelab-ua** [comment, marked as answer]
Hi @Ming-is,

Thanks for your interest in FEPX. As a note, please post general questions like this only to the Discussions forum, the Issues tracker is specifically for confirmed bugs, and feature requests that have elevated above discussions.

Concerning your question: I am not sure my answer will be satisfying. In short: on a high-fidelity representation of a microstructure, it will be very difficult to perform a simulation of 10,000 loading cycles. This, and suggestions, are detailed below.

Beginning broadly: the power of CPFEM is in its ability to provide high-fidelity predictions of the crystal-scale response of polycrystalline materials. That is, CPFEM has the ability to resolve the intra-grain deformation response via the discretization of grains into many elements, which in total compose a finite element mesh. The consequence of this is that the more elements we have in a mesh, the more computationally taxing the simulation becomes: that is, the simulation will take longer to run.

So, if you want to understand the intra-grain deformation response, and thus require fine discretization of each grain, the simulation will invariably take longer to run. This may limit the number of cycles you can presently simulate. This is alleviated somewhat by the use of parallel processing. FEPX is programmed such that it can be deployed on large-scale computational clusters, and use a large number of computational cores. This will increase the speed of simulations (to a certain point).

It is always a balance to determine the parameters of the simulations you want to perform. In your case, the most basic parameters are: the number of grains, the average number of elements per grain, the number of computational cores, and the number of loading cycles. You are often limited in one or more of these categories. In the same way that you will have a hard time performing 10,000 loading cycles, it would be difficult for a user to perform a simulation on a polycrystal with 1M grains and 1000 elements per grain.

My suggestion: Ask yourself if you need the specific results that CPFEM provides, or if another simulation method (e.g., VPSC) is more apt for your problem. If you do need CPFEM, can you reduce the size or your domain (number of grains)? Can you reduce the fidelity of the finite element mesh (number of elements per grain)? Do you have a large-scale computational cluster at your disposal? Each of these will allow the simulations to run faster. Finally, do you really need to simulate a full 10,000 cycles, or can you view trends after 10, 100, or even 1,000 cycles?

**Reply by Ming-is** [reply]
The main difficulty of my simulation is how to batch set the deformation history module in the simulation.config file. If I manually add files related to setting simulation one cycle at a time, it would be silly. After all, the contents of these files are almost duplicate. For fepx, is there any intelligent command or method that I can set the simulation file in the first few cycles, and then directly set the cycle to 10000, so that it can automatically calculate the whole cycle loading process.

**Reply by rquey** [reply]
Why not simply writing the 10000 lines to the `simulation.config` file and use`print_data` to print the results of only the cycles you need? This makes a long file, but it is only minor inconvenience.

**Reply by mkasemer** [reply]
In addition to @rquey 's suggestion: If all cycles are identical, you can also set up a simple shell script to handle this, using only one configuration file with the generic cycle defined in the load sections.

**Reply by acmelab-ua** [reply]
As promised: you can generate a shell script to handle automatic cycling.

First, a your configuration file for every cycle should just list the loading history for a single cycle. On the first cycle, you don't need to edit the configuration file. On subsequent cycles, you will want to add "restart on" to indicate to FEPX that you want to restart the simulation.

I've supplied a semi-pseduo-code below to help you in creating of your shell script:
```
for i in $(seq 001 10000) ; do
    #
    Make and enter a new simulation directory
    Copy mesh and configuration files to directory
    If on first cycle: simply run the simulation
    If on cycle greater than the first:
        Edit configuration file using `sed` to list "restart on"
        Copy previous cycle's restart files to current folder
        Run new simulation
done
```

Does this help at all?

**Comment by Ming-is** [comment]
ok.it's just what i want.Thank you very much. ------------------&nbsp;原始邮件&nbsp;------------------
***@***.***&gt;
发送时间:&nbsp;2021年7月12日(星期一) 晚上10:54
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [acmelab-ua/FEPX] 10000 cycles of cyclic loading (#26)
